### PR TITLE
[Relay, TF] Support converting TF combined_nms using Relay all_class_nms

### PR DESCRIPTION
**Pull request by masahi**
A follow up to https://github.com/apache/tvm/pull/7796 adding support for converting TF combined nms using all_class_nms. 
https://www.tensorflow.org/api_docs/python/tf/image/combined_non_max_suppression

TF and ONNX require different output format from combined NMS. To support TF, we need to return outputs differently. Coming up with the right output format for TF turned out to be non trivial, but I found the following worked quite well:

* `indices` of size (batch_size, num_boxes * num_class, 2)
* `scores` of size (batch_size, num_boxes * num_class)
* `num_detections` of size (batch_size,)

`indices` and `scores` are outputs from the NMS loop, where the selected box indices/scores come first. They are sorted within a class, but not across classes. Since TF combined NMS returns `max_total_boxes` out of `num_boxes * num_class`, the frontend code needs to do topk. 

One complication is even if `num_boxes * num_class` < `max_total_size`, TF returns `max_total_size` boxes with zero padding. Since topi/relay topk op would not work when the size of the axis to select topk elements from is less then k, we need to conditionally do topk depending on the value of `num_boxes`. Since things become complicated when `num_boxes` can be dynamic, I use `all_class_nms` code path only when `num_boxes` is static (which should be no problem in practice). Moreover, `all_class_nms` does not support `q != 1` case. So if `q != 1` or the number of boxes is dynamic, we still use the old converter.

In addition to expected speed up from parallelizing the NMS loop across different classes, this change enables running TF detection models with graph runtime, since the output from TF combined nms is always of size `max_total_size` and we no longer need to do dynamic slicing.

please review @trevor-m @anijain2305

### CODE REVIEW

File: python/tvm/relay/frontend/tensorflow.py
Comment (masahi) [inline python/tvm/relay/frontend/tensorflow.py:796]: Since I want to use this function to rewrite ONNX models, it is defined at the top level and visible from outside.

**Review by trevor-m — COMMENTED** [review]
Thanks @masahi, this is great. Some small comments.

File: python/tvm/relay/frontend/tensorflow.py
Comment (masahi) [reply python/tvm/relay/frontend/tensorflow.py:821]: oops you are right. Actually the existing converter does not seem to support both dynamic batch and num boxes either: https://github.com/apache/tvm/blob/6baccc1305b13f18ce8ac5db5fcbbadb8c278454/python/tvm/relay/frontend/tensorflow.py#L827-L828

So I'll just add an assertion to check if batch_size and num_boxes are integers and raise an error otherwise.

File: python/tvm/topi/transform.py
Comment (masahi) [reply python/tvm/topi/transform.py]: I was using this before but not anymore. Reverted.

File: python/tvm/relay/frontend/tensorflow.py
Comment (trevor-m) [inline python/tvm/relay/frontend/tensorflow.py:848]: Need to use `slice_mode="size"` for this strided_slice, otherwise using -1 for end is returning a tensor with batch size of 0.

```
%988 = vision.all_class_non_max_suppression(%986, %987, 100 /* ty=int32 */, 0.6f /* ty=float32 */, 0.3f /* ty=float32 */, meta[relay.attrs.AllClassNonMaximumSuppressionAttrs][0]) /* ty=(Tensor[(1, 172530, 2), int64], Tensor[(1, 172530), float32], Tensor[(1), int64]) */;
%989 = %988.1;
%990 = strided_slice(%989, begin=[0, 0], end=[-1, 9000], strides=[1]) /* ty=Tensor[(0, 9000), float32] */;
```

File: python/tvm/topi/cuda/nms.py
Comment (trevor-m) [inline python/tvm/topi/cuda/nms.py:1135]: Because of the int64 type, I had to set this line
https://github.com/apache/tvm/blob/main/python/tvm/topi/cuda/scan.py#L234

to `reduction[tid] = cast(0, "int64")`

Otherwise I got the error: `data type does not match content type int32 vs int64`

**Review by trevor-m — CHANGES_REQUESTED** [review]

File: python/tvm/relay/frontend/tensorflow.py
Comment (masahi) [reply python/tvm/relay/frontend/tensorflow.py:848]: I see, I'd rather use `axes` argument introduced in https://github.com/apache/tvm/pull/8165 after it is merged. This saves us from specifying the begin and end for the batch dim. I was using this in my dev branch, and realized that `axes` argument is not in `main` yet, so added this fishy workaround. And tests in `test_forward.py` have `max_output_boxes_per_class` as a variable, so we don't hit this code path.

Comment (masahi) [reply python/tvm/relay/frontend/tensorflow.py]: should be fixed

File: python/tvm/topi/cuda/nms.py
Comment (masahi) [reply python/tvm/topi/cuda/nms.py:1135]: I see. I didn't hit this error because I was not using thrust, and this code path is only for thrust. Fixed by adding `cast(0, reduction.dtype)`.

File: python/tvm/relay/frontend/tensorflow.py
Comment (trevor-m) [reply python/tvm/relay/frontend/tensorflow.py]: Thanks for fixing!

File: python/tvm/topi/cuda/nms.py
Comment (trevor-m) [reply python/tvm/topi/cuda/nms.py:1135]: Nice, thanks!

**Review by trevor-m — APPROVED** [review]
Thanks @masahi, it looks good to me now.

Initial results for TF SSD Mobilenet v1 on T4 GPU are giving me 4.56 ms. It is not quite an apples to apples comparison with the previous implementation though, so I will come back with more data soon.

**Comment by masahi** [comment]
Interesting. I'd be curious to know what it is not going faster than the previous implementation.

If you can dump the input boxes and scores to TF combined nms, I can look at the performance difference via a standalone NMS workload test. I can do that very easily for PT but no idea for TF.

**Comment by trevor-m** [comment]
The input shapes are `boxes=(1, 1917, 1, 4)` and `scores=(1, 1917, 91)`.

Here are some more numbers for TF SSD Mobilenet v1.
| Implementation | Thrust | Latency (ms) |
| --- | --- | --- |
| Previous | Yes | 3.41 |
| Previous | No | 3.85 |
| New | Yes | 4.56 |
| New | No |  2.98 ms |

It looks like without thrust this implementation is much faster!

Even if we are unable to improve the thrust performance in this PR, I still prefer this new implementation because of the static shapes. The performance is still much better than TF anyway which is ~6ms.

**Comment by masahi** [comment]
lol thrust is the bottleneck now? I see that the number of boxes is quite small in your model, I believe the new implementation would be much faster when the number of boxes is large. Do you have other models, preferably the one with more input boxes?

I've been testing tf2 models from https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf2_detection_zoo.md, their ssd mv2 model (FPNLite 320x320) has 12480 boxes,  and ssd resnet 50 v1 has more than 50000, and efficient det2 has 110484 boxes.

**Comment by masahi** [comment]
@trevor-m Also it is worth trying out the graph runtime for the new implementation, since there is no dynamic shape after NMS.

Thrust numbers actually make sense given that now we are running sort on `(batch * class, num_boxes`) along the second axis. Since batched sort with thrust is implemented via two calls to stable sort, when the axis to sort is small the overhead from the two calls become relatively big. And for small sort, TVM's sort should be faster than Thrust.

In contrast, the previous implementation does sort on `(batch, num_boxes * class)` along the second axis, so Thrust sort is fast.

**Comment by tqchen** [comment]
@trevor-m please help manage this PR, you should get merge permission once the ASF account linking is setup

**Comment by masahi** [comment]
@trevor-m you should be able to merge now.

**Comment by trevor-m** [comment]
> @trevor-m Also it is worth trying out the graph runtime for the new implementation, since there is no dynamic shape after NMS.
> 
> Thrust numbers actually make sense given that now we are running sort on `(batch * class, num_boxes`) along the second axis. Since batched sort with thrust is implemented via two calls to stable sort, when the axis to sort is small the overhead from the two calls become relatively big. And for small sort, TVM's sort should be faster than Thrust.
> 
> In contrast, the previous implementation does sort on `(batch, num_boxes * class)` along the second axis, so Thrust sort is fast.

I see, that makes sense.

I was using the models from https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf1_detection_zoo.md
I think these ones still cannot use graph runtime because there is a loop at the beginning.

**Comment by trevor-m** [comment]
Thanks @masahi !

**Pull request merged**
